Ack improvements

Matthias247 opened the pull request:
This change contains a first set of improvements to ACK handling, on the way to sending ACKs more efficiently.

In these changes the number of ACKs is actually not yet reduced, and the previous strategies are kept.
It however introduces some refactorings and prerequesites for adding ACK delay successfully:
1. The ACK state is moved into a separate struct
2. ACK delay is reported to peers, in order not to disturb RTT calculations in case ACK delay is applied
3. Pending ACKs are sent with `CONNECTION_CLOSE` frames

Ralith commented:
Are you sure the second commit is correct?

RFC9000 [§13.2.5](https://datatracker.ietf.org/doc/html/rfc9000#section-13.2.5):
> An endpoint measures the delays **intentionally** introduced between the
   time the packet with the largest packet number is received and the
   time an acknowledgment is sent.  The endpoint encodes this
   acknowledgment delay in the ACK Delay field of an ACK frame; see
   Section 19.3
(emphasis added)

Ralith commented on quinn-proto/src/connection/mod.rs:
```diff
@@ -673,28 +673,49 @@ where
+                    // TODO: Under very rare conditions the ACKs might take up so
+                    // much space that the actual close reason gets swallowed.
+                    // We could send another packet for it, but for this unlikely
+                    // condition we can also just send nothing and let the peer time out.
```
I'm not sure this is actually possible; we know we're sending at most 64 ACK ranges, and `ConnectionClose` already truncates the reason if necessary. Given the minimum MTU, I think we can rely on there being enough space for the critical parts.

Matthias247 replied:
Oh that's true. I initially had nothing, but then felt bad about potential crashes due to having no bounds check.
Probably fine to just replace the comment with "There will always be enough space"?

Ralith replied:
Sure, maybe plus a `debug_assert`.

Ralith commented on quinn-proto/src/tests/mod.rs:
```diff
@@ -930,6 +930,28 @@ fn ack_delay_report() {
+    assert!(
+        client_acks_2 > client_acks,
+        "Connection close should send pending ACKs"
+    );
```
If the client's the one sending the close, shouldn't we check the server's received ACK count?

Matthias247 replied:
Oh - yes and no - the server ACKs the clients ping, so we need to check whether the client got it.
But obviously we would need to close the server. Interesting that the test still captured the conditions :)

Matthias247 replied:
fixed that by closing the server

Matthias247 commented:
> Are you sure the second commit is correct?

Hmm, that's true. Reading that again this way it seems like we should not include this kind of delay.

Matthias247 commented:
> > Are you sure the second commit is correct?
> 
> Hmm, that's true. Reading that again this way it seems like we should not include this kind of delay.

So I thought a bit more about it:

The delay could be calculated as the time between a packet was received (`insert_one`), and the time the N-th packet unblocks sending (in `on_ack_frame_received`) or the delay timer elapsing unblocks sending a packet. In that case the the delay that is signaled would however only include the delta between the incoming packets, which might be what the spec wants us to there. For the timer elapses case its slightly different, since it's time "time between packet incoming" (on the **endpoint** task) and the time the timer elapses (on the **connection task**), so it wouldn't necessarily account for the endpoint vs connection task latency. But then again the timer would be started relative to the packet reception time, so it should end up being the same as if another packet was received during that time.

If that is done, I think `ack_delay` might be pretty fixed and repeated in retransmissions - which is probably ok since the receiver should only take it into account for the highest acknowledged PN.

Matthias247 commented:
Updated commit 2) and removed the ack delay calculation for now - it writes a static 0us. However it at least introduces the path for getting the delay from the `PendingAcks` struct and encoding the result.

djc commented on quinn-proto/src/connection/spaces.rs:
```diff
@@ -436,6 +433,72 @@ impl SendableFrames {
+            pending_acks: ArrayRangeSet::new(),
+        }
+    }
+}
```
This can be derived.

Matthias247 replied:
done

djc commented on quinn-proto/src/connection/spaces.rs:
```diff
@@ -436,6 +433,72 @@ impl SendableFrames {
+#[derive(Debug)]
+pub(crate) struct PendingAcks {
+    permit_ack_only: bool,
+    pending_acks: ArrayRangeSet,
```
Maybe rename this to `ranges` now? Feels a bit stuttery otherwise.

Matthias247 replied:
done

djc commented on quinn-proto/src/connection/mod.rs:
```diff
@@ -2747,6 +2737,40 @@ where
+    ///
+    /// This method assumes ACKs are pending, and should only be called if
+    /// `PendingAcks::can_send()` or `!PendingAcks::ranges().is_empty()` returns `true`.
+    fn populate_acks(
```
Maybe move this into `PacketSpace`? Methods without a receiver are a bit of an anti-pattern IMO, and we're always trying to move code out of `Connection` since it's already so large. I understand that means that `PacketSpace` now becomes aware of `SentFrames` and `ConnectionStats`, but that seems fine to me.

Matthias247 replied:
Although the module is definitely on the too big side, I think `populate_acks` and `populate_packet` are currently extremely similar in style and touch the same data - therefore I don't think it's better to separate them. Also `spaces.rs` currently mostly contains datastructures, and less of serialization methods.

Ralith replied:
Note that the non-receiver-ness here is just a workaround because it's called when a space is borrowed. It could be refactored to support a receiver by passing the space ID instead, and restarting the borrow afterwards in the caller, though in this instance the difference is minor.

Ralith commented on quinn-proto/src/connection/mod.rs:
```diff
@@ -2747,6 +2737,40 @@ where
+    /// Write pending ACKs into a buffer
+    ///
+    /// This method assumes ACKs are pending, and should only be called if
+    /// `PendingAcks::can_send()` or `!PendingAcks::ranges().is_empty()` returns `true`.
```
Would it be clearer to omit mention of `PendingAcks::can_send()`, since it always implies `!PendingAcks::ranges().is_empty()`?

Matthias247 replied:
Maybe - but it's debatable whether it should rather be the other way around and only `can_send()` should be checked in order to suppress always including all acks. But since that change isn't done here - and a reader of only that file might not see that `can_send` will check `!ranges().is_empy()`, I opted for putting both there. I would however be fine removing `can_send` for now if you think that's preferrable

Ralith replied:
I think `can_send` isn't actually that strong--we use that to determine if we can send an ACK *only* packet, but we still encode ACKs regardless when sending ACK-eliciting packets. Conversely, `ranges()` is always sufficient to sanely call the method, even if we don't actually want to do that under all circumstances for other reasons.

Matthias247 replied:
ok - updated the comment

Ralith commented on quinn-proto/src/connection/spaces.rs:
```diff
@@ -437,6 +437,8 @@ impl SendableFrames {
 pub(crate) struct PendingAcks {
     permit_ack_only: bool,
     ranges: ArrayRangeSet,
+    latest_incoming: Option<Instant>,
```
Let's leave a comment here explaining the intended future application, since it's otherwise mysteriously unused.

Matthias247 replied:
done

Ralith commented on quinn-proto/src/transport_parameters.rs:
```diff
@@ -139,7 +139,6 @@ impl TransportParameters {
                     .try_into()
                     .expect("setter guarantees this is in-bounds")
             }),
-            max_ack_delay: 0u32.into(),
```
Is there any reason to change this before we actually start sending nonzero values?

Matthias247 replied:
It doesn't really make a change for this anymore, and is mostly a leftover from the version where the commit reported actual delay. I decided to keep it here to minimize the size of the next change, and since it doesn't really make a difference on where it is - as long as no delay is reported the situation is the same for the peer anyway.

Plus it was kind of weird that that this field is explicitely set to something different than the default value :-)

Ralith replied:
I figured it might be useful to a peer, but realistically it's probably moot.
